File: src/utils/json/serialize/JSONSerializer.java
package utils.json.serialize;

import java.io.IOException;
import java.io.OutputStream;
import java.util.List;
import java.util.Map;

import utils.json.core.JSONArray;
import utils.json.core.JSONException;
import utils.json.core.JSONObject;

/**
 * JSONSerializer interface.
 * 
 * Defining method required to serialize JSON objects.
 * 
 * @author Tony Tsang
 */
public interface JSONSerializer {
    /**
     * Serialize a JSONObject to the specified output stream
     * 
     * @param json target json object
     * @param out output stream
     * 
     * @throws IOException
     * @throws JSONException
     */
    void serialize(JSONObject json, OutputStream out) throws IOException, JSONException;

    /**
     * Serialize a JSONObject to the specified output stream
     *
     * @param json target json object
     * @param out output stream
     * @param flushWhenFinished whether to flush when finished
     *
     * @throws IOException
     * @throws JSONException
     */
    void serialize(JSONObject json, OutputStream out, boolean flushWhenFinished) throws IOException, JSONException;
    
    /**
     * Serialize a JSONArray to the specified output stream
     * 
     * @param jsonArray target json array
     * @param out output stream
     * 
     * @throws IOException
     * @throws JSONException
     */
    void serialize(JSONArray jsonArray, OutputStream out) throws IOException, JSONException;

    /**
     * Serialize a JSONArray to the specified output stream
     *
     * @param jsonArray target json array
     * @param out output stream
     * @param flushWhenFinised whether to flush when finished
     *
     * @throws IOException
     * @throws JSONException
     */
    void serialize(JSONArray jsonArray, OutputStream out, boolean flushWhenFinised) throws IOException, JSONException;
    
    /**
     * Serialize a list, formated as JSONArray, to the specified output stream
     * 
     * @param list target list
     * @param out output stream
     * 
     * @throws IOException
     */
    void serialize(List<? extends Object> list, OutputStream out) throws IOException;

    /**
     * Serialize a list, formated as JSONArray, to the specified output stream
     *
     * @param list target list
     * @param out output stream
     * @param flushWhenFinished whether to flush when finished
     *
     * @throws IOException
     */
    void serialize(List<? extends Object> list, OutputStream out, boolean flushWhenFinished) throws IOException;
    
    /**
     * Serialize a map, formatted as JSONObject, to the specified output stream
     * 
     * @param map target map
     * @param out output stream
     *
     * @throws IOException
     */
    void serialize(Map<String, ? extends Object> map, OutputStream out) throws IOException;

    /**
     * Serialize a map, formatted as JSONObject, to the specified output stream
     *
     * @param map target map
     * @param out output stream
     * @param flushWhenFinished whether to flush when finished
     *
     * @throws IOException
     */
    void serialize(Map<String, ? extends Object> map, OutputStream out, boolean flushWhenFinished) throws IOException;

    /**
     * Serialize a JSONObject to bytes.
     *
     * @param json target json object
     * @return encoded bytes
     *
     * @throws JSONException
     * @throws IOException
     */
    byte[] serializeToBytes(JSONObject json) throws JSONException, IOException;

    /**
     * Serialize a JSONArray to bytes.
     *
     * @param jsonArray target json array
     * @return encoded bytes
     *
     * @throws JSONException
     * @throws IOException
     */
    byte[] serializeToBytes(JSONArray jsonArray) throws JSONException, IOException;

    /**
     * Serialize a map, formatted as JSONObject, to bytes.
     *
     * @param map target map
     * @return encoded bytes
     *
     * @throws IOException
     */
    byte[] serializeToBytes(Map<String, ? extends Object> map) throws IOException;

    /**
     * Serialize a list, formated as JSONArray, to bytes.
     *
     * @param list target list
     * @return encoded bytes
     *
     * @throws IOException
     */
    byte[] serializeToBytes(List<? extends Object> list) throws IOException;
}

File: src/main/kotlin/utils/json/serialize/JSONSerializer.kt
package utils.json.serialize

import utils.json.core.JSONArray
import utils.json.core.JSONObject
import java.io.IOException
import java.io.OutputStream

/**
 * JSONSerializer interface.
 * 
 * Defining method required to serialize JSON objects.
 * 
 * @author Tony Tsang
 */
interface JSONSerializer {
    /**
     * Serialize a JSONObject to the specified output stream
     * 
     * @param json target json object
     * @param out output stream
     * 
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(json: JSONObject, output: OutputStream)

    /**
     * Serialize a JSONObject to the specified output stream
     *
     * @param json target json object
     * @param out output stream
     * @param flushWhenFinished whether to flush when finished
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(json: JSONObject, output: OutputStream, flushWhenFinished: Boolean)
    
    /**
     * Serialize a JSONArray to the specified output stream
     * 
     * @param jsonArray target json array
     * @param out output stream
     * 
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(jsonArray: JSONArray, output: OutputStream)

    /**
     * Serialize a JSONArray to the specified output stream
     *
     * @param jsonArray target json array
     * @param out output stream
     * @param flushWhenFinised whether to flush when finished
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(jsonArray: JSONArray, output: OutputStream, flushWhenFinished: Boolean)
    
    /**
     * Serialize a list, formated as JSONArray, to the specified output stream
     * 
     * @param list target list
     * @param out output stream
     * 
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(list: List<Any?>, output: OutputStream)

    /**
     * Serialize a list, formated as JSONArray, to the specified output stream
     *
     * @param list target list
     * @param out output stream
     * @param flushWhenFinished whether to flush when finished
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(list: List<Any?>, output: OutputStream, flushWhenFinished: Boolean)
    
    /**
     * Serialize a map, formatted as JSONObject, to the specified output stream
     * 
     * @param map target map
     * @param out output stream
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(map: Map<String, Any?>, output: OutputStream)

    /**
     * Serialize a map, formatted as JSONObject, to the specified output stream
     *
     * @param map target map
     * @param out output stream
     * @param flushWhenFinished whether to flush when finished
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serialize(map: Map<String, Any?>, output: OutputStream, flushWhenFinished: Boolean)

    /**
     * Serialize a map object to the specified output stream
     * 
     * @param mapObject map object
     * @param output output stream
     */
    @Throws(IOException::class)
    fun serialize(mapObject: Any, output: OutputStream)

    /**
     * Serialize a MapObject to the specified output stream
     * 
     * @param mapObject map object
     * @param output output stream
     * @param flushWhenFinished whether to flush when finished
     */
    @Throws(IOException::class)
    fun serialize(mapObject: Any, output: OutputStream, flushWhenFinished: Boolean)

    /**
     * Serialize a JSONObject to bytes.
     *
     * @param json target json object
     * @return encoded bytes
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serializeToBytes(json: JSONObject): ByteArray

    /**
     * Serialize a JSONArray to bytes.
     *
     * @param jsonArray target json array
     * @return encoded bytes
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serializeToBytes(jsonArray: JSONArray): ByteArray

    /**
     * Serialize a map, formatted as JSONObject, to bytes.
     *
     * @param map target map
     * @return encoded bytes
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serializeToBytes(map: Map<String, Any?>): ByteArray

    /**
     * Serialize a list, formated as JSONArray, to bytes.
     *
     * @param list target list
     * @return encoded bytes
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun serializeToBytes(list: List<Any?>): ByteArray

    /**
     * Serialize a MapObject to bytes
     * 
     * @param mapObject map object
     */
    fun serializeToBytes(mapObject: Any): ByteArray
}
